### ISSUE TITLE
[FEA] desloc_bench: 梯度压缩率 vs 收敛影响

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#583]
## Summary

测量不同梯度压缩率（Top-K 稀疏化）对 DES-LOC 训练收敛的影响，给出"收敛损失 ≤ 5%" 的最大安全压缩率推荐值。

## Design Reference

- **File**: `benchmarks/desloc_compression_rate.py` (新建)
- **Parent EPIC**: #299 | **Parent THEME**: #276

## Background

DES-LOC 在 tier 间同步慢动量时默认对梯度做 Top-K 稀疏化以节省跨 PCIe/NVLink 带宽。压缩率越高（保留比例越低），通信量越小，但梯度信息损失越大，可能导致收敛变慢或最终 loss 偏高。

本 benchmark 系统地扫描压缩率 [1%, 5%, 10%, 50%, 100%]，在小型验证任务（GPT-125M + WikiText-103 子集）上对比各设置下的训练曲线，给出兼顾通信效率与收敛质量的工程推荐。

## Implementation

```python
# benchmarks/desloc_compression_rate.py
"""
Gradient compression rate vs convergence benchmark.

Compression rates: 1%, 5%, 10%, 50%, 100% (no compression)
Model: GPT-125M
Dataset: WikiText-103 first 100K tokens (fast proxy)
Steps: 1000 (sufficient to see convergence trend)
Metric: eval loss at step {200, 400, 600, 800, 1000}
"""

import json
import math
import time
from pathlib import Path
from typing import Iterator

import torch
import torch.distributed as dist
from deepspeed.core.hetero_bridge import DeslocEngine, CompressionConfig
from deepspeed.core.transformer import TransformerConfig, GPTModel


COMPRESSION_RATES = [0.01, 0.05, 0.10, 0.50, 1.00]
EVAL_STEPS = [200, 400, 600, 800, 1000]


def build_model(rank: int) -> GPTModel:
    cfg = TransformerConfig(
        num_layers=12, hidden_size=768, num_attention_heads=12,
        ffn_hidden_size=3072, vocab_size=50257, seq_length=512,
    )
    return GPTModel(cfg).to(rank)


def load_wikitext_tokens(path: str = "data/wikitext103_first100k.pt") -> torch.Tensor:
    """Pre-tokenized WikiText-103 subset, shape (N,)."""
    return torch.load(path)


def token_batches(tokens: torch.Tensor, batch_size: int = 8,
                  seq_len: int = 512) -> Iterator[torch.Tensor]:
    """Yield (batch_size, seq_len) batches indefinitely."""
    n = (len(tokens) - 1) // seq_len
    idx = 0
    while True:
        batch = []
        for _ in range(batch_size):
            start = (idx % n) * seq_len
            batch.append(tokens[start: start + seq_len])
            idx += 1
        yield torch.stack(batch)


@torch.no_grad()
def eval_loss(engine, tokens: torch.Tensor, rank: int,
              n_batches: int = 20) -> float:
    engine.eval()
    losses = []
    batches = token_batches(tokens)
    for _ in range(n_batches):
        x = next(batches).to(rank)
        loss = engine(x, labels=x).loss
        losses.append(loss.item())
    engine.train()
    return sum(losses) / len(losses)


def run_single_rate(rank: int, world_size: int, compression_rate: float,
                    tokens: torch.Tensor, total_steps: int = 1000) -> dict:
    model = build_model(rank)
    comp_cfg = CompressionConfig(topk_ratio=compression_rate)

    engine = DeslocEngine(
        model=model,
        rank=rank,
        tiers=[
            {"ranks": [0, 1], "device_type": "H100", "mem_gb": 80},
            {"ranks": [2, 3, 4], "device_type": "A6000", "mem_gb": 48},
        ],
        K=5,
        slow_momentum=0.9,
        compression=comp_cfg,
    )
    optimizer = torch.optim.AdamW(model.parameters(), lr=2e-4, weight_decay=0.1)
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=total_steps)

    history = {"compression_rate": compression_rate, "eval_steps": {}}
    t0 = time.perf_counter()
    data_iter = token_batches(tokens)

    for step in range(1, total_steps + 1):
        x = next(data_iter).to(rank)
        loss = engine(x, labels=x).loss
        loss.backward()
        optimizer.step()
        scheduler.step()
        optimizer.zero_grad()

        if step in EVAL_STEPS:
            val_loss = eval_loss(engine, tokens, rank)
            if rank == 0:
                ppl = math.exp(val_loss)
                history["eval_steps"][step] = {"loss": round(val_loss, 4),
                                                "ppl": round(ppl, 2)}
                print(f"[rate={compression_rate:.0%} step={step}] "
                      f"loss={val_loss:.4f} ppl={ppl:.2f}")

    history["total_time_s"] = round(time.perf_counter() - t0, 1)
    return history


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--rate", type=float, choices=COMPRESSION_RATES,
                        help="Single rate to run (omit to run all)")
    parser.add_argument("--steps", type=int, default=1000)
    parser.add_argument("--data", default="data/wikitext103_first100k.pt")
    parser.add_argument("--output", default="results/compression_rate.jsonl")
    args = parser.parse_args()

    dist.init_process_group("nccl")
    rank = dist.get_rank()
    world_size = dist.get_world_size()

    tokens = load_wikitext_tokens(args.data)
    rates = [args.rate] if args.rate else COMPRESSION_RATES

    for rate in rates:
        dist.barrier()
        result = run_single_rate(rank, world_size, rate, tokens, args.steps)
        if rank == 0:
            # Compute degradation vs no-compression baseline
            out = args.output
            Path(out).parent.mkdir(parents=True, exist_ok=True)
            with open(out, "a") as f:
                f.write(json.dumps(result) + "
")

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

### 运行命令

```bash
# 逐率运行 (推荐分开跑，方便 resume)
for RATE in 0.01 0.05 0.10 0.50 1.00; do
  torchrun --nproc_per_node=5 benchmarks/desloc_compression_rate.py \
    --rate $RATE --steps 1000
done

# 合并结果并生成对比表
python3 scripts/summarize_compression.py results/compression_rate.jsonl
```

### 预期结果格式

```json
{"compression_rate": 0.05, "eval_steps": {
  "200": {"loss": 4.312, "ppl": 74.5},
  "1000": {"loss": 3.871, "ppl": 47.9}
}, "total_time_s": 892.3}
```

### 收敛对比示意

```
Rate   | Final Loss | Final PPL | vs No-Compress | Time (s)
-------|-----------|-----------|----------------|--------
1%     |   3.94    |   51.4    |   +2.3%        |  821
5%     |   3.88    |   48.3    |   +0.8%        |  848
10%    |   3.87    |   47.9    |   +0.5%        |  859
50%    |   3.86    |   47.5    |   +0.3%        |  891
100%   |   3.85    |   47.1    |   baseline     |  953

推荐: 5%–10% 压缩率（收敛损失 <1%，通信量减少 90%+）
```

## Acceptance Criteria

- [ ] 压缩率: 1%, 5%, 10%, 50%, 100%
- [ ] 收敛: 各压缩率下的最终 loss（step 1000）及中间检查点
- [ ] 推荐: 最佳压缩率（收敛 ≤ 5% 损失），写入 `results/compression_rate.jsonl`
- [ ] 支持 `--rate` 单率运行（可 resume）
- [ ] rank=0 实时打印 loss/ppl

## Related

- #581 K 值灵敏度分析
- #585 Ablation 各组件贡献
- #299 EPIC: desloc_bench
